### ISSUE TITLE
Knex seed for postgres

### ISSUE THREAD

**Issue by mairh** [sysgears/apollo-fullstack-starter-kit#187]
I was playing around with different databases for this kit since we are mentioning over [here](https://github.com/sysgears/apollo-fullstack-starter-kit/blob/master/README.md#features) that any database can be plugged into the kit in the same way as SQLite.

So, I tried plugging `postgres` and faced an issue with the knex seed provided in this kit. The problem is we are using `truncate()` for seed migration and `comment` table is referencing `post` table to populate the comments for a particular `post_id`.

Since `postgres` does not allow truncating tables if the table has any foreign key reference making the script `yarn seed` fails.

![screen shot 2017-05-17 at 15 00 54](https://cloud.githubusercontent.com/assets/4072250/26152901/afd9d87c-3b11-11e7-8958-95ef653f767c.png)

However, using raw queries from knex does push the seed to completion, but it does not insert the `post_id` reference in `comment` table.

From this (in seeds/post.js)
```
await Promise.all([
    knex('post').truncate(),
    knex('comment').truncate()
]);
```

to this
```
await Promise.all([
    knex.raw('TRUNCATE TABLE post CASCADE'),
    knex.raw('TRUNCATE TABLE comment CASCADE')
]);
```

However, it returns `null` for `post_id`

![screen shot 2017-05-17 at 15 05 31](https://cloud.githubusercontent.com/assets/4072250/26153146/cff860be-3b12-11e7-86b6-6df33e31e5f7.png)

This is the final seed function in case of postgres

```
export async function seed(knex, Promise) {
  await Promise.all([
    knex.raw('TRUNCATE TABLE post CASCADE'),
    knex.raw('TRUNCATE TABLE comment CASCADE')
  ]);

  await Promise.all([...Array(20).keys()].map(async ii => {
    const post = await knex('post').insert({
      title: `Post title ${ii + 1}`,
      content: `Post content ${ii + 1}`
    });

    await Promise.all([...Array(2).keys()].map(async jj => {
      return knex('comment').insert({
        post_id: post[0],
        content: `Comment title ${jj + 1} for post ${post[0]}`
      });
    }));
  }));
}
```

P.S -> Just in case you want to see my modified knexdata.js file

```
export const development = {
  client: 'pg',
  debug: true,
  connection: {
    host: '127.0.0.1',
    user: 'postgres',
    password: '*****',
    database: 'dev_db'
  },
  seeds: {
    directory: './src/server/database/seeds'
  },
  migrations: {
    directory: './src/server/database/migrations'
  },
  useNullAsDefault: true
};
```

Another problem is the primary key (autoincrement) for `post` and `comment` table is not resetting back to `1` after running the seed every time.

![screen shot 2017-05-17 at 15 16 20](https://cloud.githubusercontent.com/assets/4072250/26153633/eda602e0-3b14-11e7-9ec1-698fdc0a750e.png)

Is there any workaround for this?

**Comment by bagshik** [comment]
Hey, I was in the same situation as you. I did something similar in the seed file :

```export async function seed(knex, Promise) {
  await Promise.all([
    knex.raw('ALTER SEQUENCE post_id_seq RESTART WITH 1'),
    knex.raw('ALTER SEQUENCE comment_id_seq RESTART WITH 1'),
    knex.raw('TRUNCATE TABLE post, comment CASCADE'),
    knex.raw('TRUNCATE TABLE comment, post CASCADE')
  ]);

  await Promise.all([...Array(20).keys()].map(async ii => {
    const post = await knex('post').returning('id').insert({
      title: `Post title ${ii + 1}`,
      content: `Post content ${ii + 1}`
    });

    await Promise.all([...Array(2).keys()].map(async jj => {
      return knex('comment').insert({
        post_id: post[0],
        content: `Comment title ${jj + 1} for post ${post[0]}`
      });
    }));
  }));
}
```

This will make sure to reset the autoincrement and populate the post_id fields for the comments. Let me know if this works for you.

PS : I still have a weird issue when I'm running the seed files, the `id` of the posts doesn't match its `title` and `content` fields, did you also run into this problem?

**Comment by mairh** [comment]
@bagshik Thanks. That did the trick. I will try to find a neat solution (if any).

Regarding your problem with `id` of the post is not matching its `title` and `content` is because you are running the truncation for `post` and `comment` table twice.

Try this in sequential promises
```
await Promise.all([
    knex.raw('ALTER SEQUENCE post_id_seq RESTART WITH 1'),
    knex.raw('ALTER SEQUENCE comment_id_seq RESTART WITH 1'),
    knex.raw('TRUNCATE TABLE post CASCADE'),
    knex.raw('TRUNCATE TABLE comment CASCADE')
]);
```

**Comment by mairh** [comment]
@bagshik Yes, that did the trick. It would be nice to find a neat solution though (maybe that's the only one at the moment).

Due to the `async` operations, I do see the `id` field is not matching the `title` and `content` fields

**Comment by bagshik** [comment]
@mairh Yeah, I agree. I'm still looking though, I'll let you know if I find a better solution.

**Comment by mairh** [comment]
I'll keep the issue open and wait for @vlasenko insight.

**Comment by mairh** [comment]
@bagshik By the way why are you truncating `post` and `comment` table twice.

Only running `knex.raw('TRUNCATE TABLE post, comment CASCADE')` should be enough.

**Comment by mairh** [comment]
@bagshik Does adding a new post or comment works for you while using postgress, and with the above configuration?

I am getting the following error when adding post and comment

Adding Post
<img width="1246" alt="screen shot 2017-05-17 at 17 29 07" src="https://cloud.githubusercontent.com/assets/4072250/26159152/7768c0ce-3b26-11e7-9170-ed482555ed2c.png">

Adding Comment
<img width="1265" alt="screen shot 2017-05-17 at 17 29 38" src="https://cloud.githubusercontent.com/assets/4072250/26159167/8194f694-3b26-11e7-9ce8-6741b2ebd791.png">

**Comment by bagshik** [comment]
@mairh Oh I didn't notice this, you're right about that. I just tried adding a new post / comment, I'm getting the same errors as you but the data is still getting inserted into the database. We'll need to look into this... I'll let you know if I find a solution.

**Comment by vlasenko** [comment]
`Undefined binding(s)` error when adding post should be fixed via:
https://github.com/sysgears/apollo-fullstack-starter-kit/commit/e3406ae46dc18774b41121151a99bd023a03cc23

**Comment by mairh** [comment]
Thanks. I will check it out and close the issue after some testing tonight.

**Comment by mairh** [comment]
This issue can be closed once https://github.com/sysgears/apollo-fullstack-starter-kit/pull/188 is merged.

**Comment by vlasenko** [comment]
Fixed by #188
